### ISSUE TITLE
Notification event types mismatches

### ISSUE THREAD

**Issue by maxceem** [appirio-tech/connect-app#3508]
Follow up from https://github.com/appirio-tech/connect-app/issues/3375#issuecomment-577492847

During creating a list of notifications which we are not showing in Connect App intentionally and we want to ignore it's appeared that some notifications have mismatches in the event type naming and shouldn't be ignored, but should be fixed instead in some services to match:

![image](https://user-images.githubusercontent.com/146016/72955432-e4d9dc00-3dd6-11ea-8041-b3d1b6b84d78.png)

**Comment by vikasrohit** [comment]
Just to clarify @maxceem, `added to be ignored` are notifications which should not be ignored, but are in ignore list right now, correct?

**Comment by maxceem** [comment]
Not sure what you mean. Let me rephrase.

- Initially, we were showing warnings for all notifications that come to the Connect App but are not shown.
- As we intentionally didn't want to show them we decided to create a ignore list, so we don't show warnings for them. While if someday we have some new unsupported notification type, we still want to see a warning to make a decision: show it or add to the ignore list. So warnings should show us actual situations where we should take some action.
- This has been implemented in this issue https://github.com/appirio-tech/connect-app/issues/3375
- Though during the review, I've noticed, that most likely these 3 notifications shouldn't be ignored, and the warnings are showing valid cases where something is wrong with notification types. Most likely notification types mismatch between services and should be fixed somewhere to match.

**Comment by vikasrohit** [comment]
> Though during the review, I've noticed, that most likely these 3 notifications shouldn't be ignored, and the warnings are showing valid cases where something is wrong with notification types. Most likely notification types mismatch between services and should be fixed somewhere to match.

Got it.

**Comment by maxceem** [comment]
Looks like in Connect App these events indeed have the wrong names. They should be removed from ignored, and we should fix the names instead.

In Project Service we initially create these events: https://github.com/topcoder-platform/tc-project-service/blob/develop/src/constants.js#L220-L234

![image](https://user-images.githubusercontent.com/146016/73507278-78349200-4413-11ea-9a92-139cf8eef04c.png)

And in Tc Notifications Service we continue using these names: https://github.com/topcoder-platform/tc-notifications/blob/dev/connect/constants.js#L59-L62

![image](https://user-images.githubusercontent.com/146016/73507301-8a163500-4413-11ea-9615-754e1e266732.png)

But in Connect App, currently, names don't match:
https://github.com/appirio-tech/connect-app/blob/dev/src/config/constants.js#L941-L944

![image](https://user-images.githubusercontent.com/146016/73507235-4fac9800-4413-11ea-80e0-744fda3b948f.png)

**Comment by maxceem** [comment]
How to test:
*(you would need two browsers with 2 different users)*
1. Create a new project with one Manager user, we would call him **ManagerReceiver**
2. As a **ManagerReceiver** open Project Listing page (go out from the project page to avoid notifications marked as read automatically)
3. Join project with another Manager - **ManagerTrigger**
4. As a **ManagerTrigger** change some Phase status to `In Progress`:

    ![image](https://user-images.githubusercontent.com/146016/73508580-589f6880-4417-11ea-9c34-0d2efe499fa0.png)

5. As a  **ManagerReceiver** reload the page and observe that in notification dropdown there is `updated the phase timeline`:

    ![image](https://user-images.githubusercontent.com/146016/73508642-8be1f780-4417-11ea-8e34-bb46d9d62f8d.png)

   - ✅`connect.notification.project.timeline.adjusted`

6. As a **ManagerTrigger** click **Mark as Completed** to the milestone of the Phase:

   ![image](https://user-images.githubusercontent.com/146016/73508718-ccda0c00-4417-11ea-8db5-3b849ebfd80c.png)

7. As a **ManagerReceiver** reload the page and observe that in notification dropdown there are `activated a milestone in the phase` and `completed a milestone in the phase`:

    ![image](https://user-images.githubusercontent.com/146016/73510738-2d6c4780-441e-11ea-9e55-b6c303fc290b.png)

    - ✅`connect.notification.project.timeline.milestone.transition.active`
    - ✅`connect.notification.project.timeline.milestone.transition.completed`

8. As a **ManagerTrigger** continue clicking **Mark as Completed** until click it for some phase **Delivery** milestone so you don't see such a button anymore:
 
    ![image](https://user-images.githubusercontent.com/146016/73511083-65c05580-441f-11ea-973e-4cd48aa8a1dc.png)

7. As a **ManagerReceiver** reload the page and observe that in notification dropdown there is `waiting for customer on a milestone in...`:

    ![image](https://user-images.githubusercontent.com/146016/73511125-91dbd680-441f-11ea-840d-7b2ffc13b08d.png)

    - ✅`connect.notification.project.timeline.milestone.waiting.customer`

**Comment by vikasrohit** [comment]
@maxceem aren't we using new events names as per v5 standards https://github.com/topcoder-platform/tc-project-service/blob/develop/src/constants.js#L152-L158? 

Further, in the last post by you, I didn't see any missing notifications. I guess, you are conveying that these all events are now fixed? Am I correct?

**Comment by maxceem** [comment]
@vikasrohit 

> @maxceem aren't we using new events names as per v5 standards https://github.com/topcoder-platform/tc-project-service/blob/develop/src/constants.js#L152-L158?

We use new standard notification events for handling indexing data in ES via new `project-processor-es`. But for the **TC Notification** service we keep our old Kafka events for now. If we want to use new events, we should come with an idea of where we should put the logic to process them. As if we just send new events to **TC Notification** service, then **TC Notification** should know about the internal structure of Project Service data to generate the same notifications as we have now.

> Further, in the last post by you, I didn't see any missing notifications. I guess, you are conveying that these all events are now fixed? Am I correct?

They are fixed now in DEV. Though they are still missing on PROD, I can see that they are listed in the console as ignored on PROD:

![image](https://user-images.githubusercontent.com/146016/73639210-de851300-46a6-11ea-8ed6-b75c2563ceaa.png)

And there are no notifications in the notification dropdown like:
- `updated the phase timeline`
- `activated a milestone in the phase`
- `completed a milestone in the phase`
- `waiting for customer on a milestone in...`

**Comment by vikasrohit** [comment]
> TC Notification should know about the internal structure of Project Service data to generate the same notifications as we have now.

I think tc-notifications is already aware of the payload structure. I guess, even in the latest [attempt](https://github.com/topcoder-platform/tc-notifications/tree/dev/src/processors) to make the general consumers, it is coupled with calling application. What I am thinking of to make it more loosely coupled is to use config like [this](https://github.com/topcoder-platform/tc-notifications/blob/dev/config/default.js#L48)  but make it dynamic by storing it in database and let the caller store the configuration in database and then have general code that uses config defined to generate the specific notifications. This might use our expression evaluator to turn on/off common features (e.g. generating notification for the set of users) supported by the notifications processor. However, I think that is somewhat ambitious task for now and we might not be able to tackle as of now. So, I would like to proceed with consuming the new common events with existing code base, e.g. we can add a common lister of all project.create events and then fan out specific events to specific handlers.

**Comment by nithyaasworld** [comment]
@maxceem On Step:5 [here](https://github.com/appirio-tech/connect-app/issues/3508#issuecomment-580569330), its failing for me.

That is, I am not receiving any notification for any of the phase updates. I double checked that:

- _ManagerReceiver_ is not on the project details page, to avoid auto-marking notifications as read
- _Manager Receiver_ has all notifications enabled in notification settings

My _ManagerReceiver_ User: TCConManager
My _ManagerTrigger_ User: TCConAdmin
The Project I tested with: https://connect.topcoder-dev.com/projects/8851

Every time I refresh the page, **all I see in console is some CORS errors.** Could this be the reason for not receiving the notification? 


Video: 
[notification CORS issue.zip](https://github.com/appirio-tech/connect-app/files/4150043/notification.CORS.issue.zip)

**Comment by maxceem** [comment]
@nithyaasworld it was due to some other branch deployed for TC Notification Service by another team, so there were no notifications at all.

Now TC Notification Service works again on DEV, could you please try now.

**Comment by nithyaasworld** [comment]
@maxceem I am still seeing the same issue as described earlier in Dev env. Can you please check again? Thanks.

**Comment by nithyaasworld** [comment]
I checked for the same users and same project as earlier. Same CORS errors are seen and No notifications received.

**Comment by maxceem** [comment]
Ah, I've got, **Connect Manager** doesn't get such notifications. Please, use **Manager** user a receiver.

**Comment by maxceem** [comment]
@nithyaasworld, please ignore the above, **Connect Manager** should get notifications.

**Comment by maxceem** [comment]
@nithyaasworld it was some issue with TC Notification Service on DEV not relevant to any of our changes so service was partially down. Could you please try to test this again.

**Comment by vikasrohit** [comment]
Closing. Please feel free to create new issues for any regression.
